File: 프로그래머스/Level 2/괄호 회전하기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/76502
def check(s):
    stack = []
    for ss in s:
        if ss == "(" or ss == "[" or ss == "{":
            stack.append(ss)
        else:
            if not stack:
                return False
            elif ss == ")" and stack[-1] == "(":
                stack.pop()
            elif ss == "]" and stack[-1] == "[":
                stack.pop()
            elif ss == "}" and stack[-1] == "{":
                stack.pop()
            else:
                return False
    if stack:
        return False
    return True


def solution(s):
    answer = 0
    s = list(s)

    for _ in range(1, len(s)):  # x (0 ≤ x < (s의 길이)) 칸만큼 회전
        if check(s):  # 올바른 괄호
            answer += 1

        # 회전
        s.append(s.pop(0))
    return answer

File: 프로그래머스/Level 2/괄호 회전하기.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/76502
import Foundation

func solution(_ s:String) -> Int {
    var answer:Int = 0
    var sArr = s.map{ String($0) }
    
    func isCorrect(_ sArr: [String]) -> Bool {
        var stack:[String] = []
        
        for ch in sArr {
            if ch == "[" || ch == "(" || ch == "{" {
                stack.append(ch)
            } else {
                if stack.isEmpty {
                    return false
                } else {
                    if stack.last! == "[" && ch == "]" {
                        stack.removeLast()
                    } else if stack.last! == "(" && ch == ")" {
                        stack.removeLast()
                    } else if stack.last! == "{" && ch == "}" {
                        stack.removeLast()
                    }
                }
            }
        }
        return stack.isEmpty ? true : false 
    }
    
    for i in 0..<s.count {
        if isCorrect(sArr) {
            answer += 1
        }
        sArr.append(sArr.removeFirst())
    }
    
    return answer
}